src/Connect/buffer.hpp

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+// The path length constants live in gui :-(
+#include <gui/file_list_defs.h>
+
+#include <array>
+#include <optional>
+#include <cassert>
+#include <cstdint>
+#include <memory>
+
+namespace con {
+
+template <size_t S>
+class Buffer {
+private:
+    std::array<uint8_t, S> data;
+    bool borrowed = false;
+
+public:
+    class Borrow {
+    private:
+        Buffer *buff;
+        friend class Buffer<S>;
+        Borrow(Buffer *buff)
+            : buff(buff) {}
+
+    public:
+        Borrow(const Borrow &) = delete;
+        Borrow &operator=(const Borrow &) = delete;
+        Borrow(Borrow &&other)
+            : buff(other.buff) {
+            other.buff = nullptr;
+        }
+        Borrow &operator=(Borrow &&other) {
+            if (buff != nullptr) {
+                assert(buff->borrowed);
```

Comment:
Where are all these aserts implemented? Are we sure this resoults in some form of error screen?

Reply:
I'm 100% sure that asserts:

* Don't result in any kind of error screen, the thread (unfortunately, only that one thread) just stops. That's somewhat less useful than what it could do, but it does make the developer go and investigate and the debugger points very clearly at the culprit (unlike any kind of `bsod`, which just resets the printer and loses all the valuable stack trace, so `bsod` is even less useful).
* Don't get into release builds. They are only part of the debug builds to help catch bugs while implementing things.

My primary intention with using asserts is, however, documentation of the intention. It's more useful (and often more precise in the information) than a comment. Given there's nothing better right now, I use at least asserts that are sub-optimal then just not have anything at all. @dragomirecky mentioned he would like to make asserts more useful in the future in some way, but I'm not sure if there was any progress or any concrete plan to do so.

Reply:
If that is the intention, it might be useful to invoke SW breakpoint in the assert, but that's beyond the scope of this PR.
Just a thought for future improvements.